File: core/src/main/scala/tripaint/effects/CellEffect.scala
package tripaint.effects

import tripaint.{Color, ColorLookup}
import tripaint.coords.GlobalPixCoords

class CellEffect extends LocalEffect {
  def name: String = "Cell"

  override protected def predicate(
      image: ColorLookup,
      here: GlobalPixCoords
  )(coords: GlobalPixCoords, color: Color): Boolean = {
    coords.cell == here.cell
  }

  override protected def weightedColor(image: ColorLookup, here: GlobalPixCoords)(
      coords: GlobalPixCoords
  ): (Double, Color) = {
    (1, image.lookup(coords).get)
  }
}

File: core/src/main/kotlin/tripaint/effects/CellEffect.kt
package tripaint.effects

import tripaint.ColorLookup
import tripaint.color.Color
import tripaint.coords.GlobalPixCoords

class CellEffect : LocalEffect() {
    override fun name(): String = "Cell"

    override fun predicate(
        image: ColorLookup,
        here: GlobalPixCoords
    ): (coords: GlobalPixCoords, color: Color) -> Boolean {
        return { coords, _ -> coords.cell() == here.cell() }
    }

    override fun weightedColor(image: ColorLookup, here: GlobalPixCoords): (coords: GlobalPixCoords) -> Pair<Double, Color> {
        return { coords -> Pair(1.0, image.lookup(coords)!!) }
    }
}
